### PR TITLE
Implement inequality operators (<, >, <=, >=) in the enable when statement

### LINKED ISSUE #818: Implement inequality operators (<, >, <=, >=) in the enable when statement
**Is your feature request related to a problem? Please describe.**
See documentation for enable when operators: https://www.hl7.org/fhir/valueset-questionnaire-enable-operator.html

At the moment we support exists/eq/not eq but none of the inequality operators.

**Describe the solution you'd like**
Code change needs to be made here: https://github.com/google/android-fhir/blob/7351fecee6dea9b14ba6c073e09ff806d73438f9/datacapture/src/main/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluator.kt#L120

to support the 4 operators missing.

**Describe alternatives you've considered**
Don't think we can use things like fhir path engine. this needs to be implemented in the library.

**Additional context**
NA

**Would you like to work on the issue?**
NA

### PR DESCRIPTION
**Pull request by santosh-pingle**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #818 

**Description**
Implement inequality operators (<, >, <=, >=) in the enable when statement

**Alternative(s) considered**
no

**Type**
Choose one: Bug fix

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md)
- [x] I have read [How to Contribute](https://github.com/google/android-fhir/blob/master/docs/contributing.md)
- [x] I have read the [Developer's guide](https://github.com/google/android-fhir/wiki/Developer's-Guide)
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate )
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally
- [x] I have built and run the reference app(s) to verify my change fixes the issue and/or does not break the reference app(s)

### CODE REVIEW

**Comment by jingtang10** [comment]
please update this PR now that #847 is merged

**Review by jingtang10 — CHANGES_REQUESTED** [review]
please udpate

**Comment by santosh-pingle** [comment]
Updating PR.

**Comment by santosh-pingle** [comment]
@jingtang10 
Can you please review it?
I will check whether I can add more tests to increase code coverage.

**Review by jingtang10 — CHANGES_REQUESTED** [review]
as discussed, merge the `MoreTypes.kt` files and move it to the `common` module and use the `Type.compareTo` rather than defining the extension functions on the `QuestionnaireItemEnableWhenComponent`.

**Comment by santosh-pingle** [comment]
Working on review comments.

**Comment by santosh-pingle** [comment]
#923

**Comment by santosh-pingle** [comment]
> as discussed, merge the `MoreTypes.kt` files and move it to the `common` module and use the `Type.compareTo` rather than defining the extension functions on the `QuestionnaireItemEnableWhenComponent`.

done.

**Review by jingtang10 — CHANGES_REQUESTED** [review]
In general we shouldn't be making any code change without unit test change.

If `MoreTypes.kt` file is changed please add unit tests as well.

I don't believe the quantity type needs to be singled out in the unit test for the enablement evaluator -- that's an irrelevant detail in the code of enablement evaluator. but that's a relevant detail in the `MoreTypes.kt` file, so the tests should reside there.

**Comment by santosh-pingle** [comment]
yes, quantity tests need to be moved to ```MoreTypes.kt```, thanks!

File: common/src/main/java/com/google/android/fhir/MoreTypes.kt [hunk 1]
```diff
@@ -62,6 +63,18 @@ operator fun Type.compareTo(value: Type?): Int {
       this.fhirType().equals("dateTime") -> {
         return this.dateTimeValue().value.compareTo(value.dateTimeValue().value)
       }
+      this.fhirType().equals("Quantity") -> {
```
Comment (jingtang10) [inline common/src/main/java/com/google/android/fhir/MoreTypes.kt:66]: can you try to change the function signature to this:

`operator fun Type.compareTo(value: Type): Int {`

i.e. remove the `?` in the parameter type.

and then remove line 46 as a check? thanks

File: datacapture/src/test/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluatorTest.kt [hunk 2]
```diff
@@ -383,6 +383,110 @@ class EnablementEvaluatorTest {
...
+          operator = Questionnaire.QuestionnaireItemOperator.GREATER_THAN,
+          expected = IntegerType(10),
+          actual = listOf(IntegerType(5))
+        )
+      )
+      .isFalse()
+  }
+
+  @Test
+  fun evaluate_greaterOrEqual_answersGreaterOrEqualEnableWhenAnswer_shouldReturnTrue() {
```
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluatorTest.kt:413]:
```suggestion
  fun evaluate_expectAnswerGreaterThanOrEqualToValue_answerGreaterThanOrEqualToValue_shouldReturnTrue() {
```

File: datacapture/src/test/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluatorTest.kt [hunk 3]
```diff
@@ -383,6 +383,110 @@ class EnablementEvaluatorTest {
...
+          operator = Questionnaire.QuestionnaireItemOperator.LESS_THAN,
+          expected = IntegerType(10),
+          actual = listOf(IntegerType(20))
+        )
+      )
+      .isFalse()
+  }
+
+  @Test
+  fun evaluate_lessOrEqual_answerLessOrEqualEnableWhenAnswer_shouldReturnTrue() {
```
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluatorTest.kt:465]:
```suggestion
  fun evaluate_expectAnswerLessThanOrEqualToValue_someAnswerLessThanOrEqualToValue_shouldReturnTrue() {
```

File: datacapture/src/test/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluatorTest.kt [hunk 4]
```diff
@@ -383,6 +383,110 @@ class EnablementEvaluatorTest {
...
+          operator = Questionnaire.QuestionnaireItemOperator.LESS_OR_EQUAL,
+          expected = IntegerType(10),
+          actual = listOf(IntegerType(10))
+        )
+      )
+      .isTrue()
+  }
+
+  @Test
+  fun evaluate_lessOrEqual_noAnswerLessOrEqualEnableWhenAnswer_shouldReturnFalse() {
```
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluatorTest.kt:478]:
```suggestion
  fun evaluate_expectAnswerLessThanOrEqualToValue_noAnswerLessThanOrEqualToValue_shouldReturnFalse() {
```

File: datacapture/src/test/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluatorTest.kt [hunk 5]
```diff
@@ -383,6 +383,110 @@ class EnablementEvaluatorTest {
...
+          operator = Questionnaire.QuestionnaireItemOperator.GREATER_OR_EQUAL,
+          expected = IntegerType(10),
+          actual = listOf(IntegerType(10))
+        )
+      )
+      .isTrue()
+  }
+
+  @Test
+  fun evaluate_greaterOrEqual_noAnswersGreaterOrEqualEnableWhenAnswer_shouldReturnFalse() {
```
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluatorTest.kt:426]:
```suggestion
  fun evaluate_expectAnswerGreaterThanOrEqualToValue_noAnswerGreaterThanOrEqualToValue_shouldReturnFalse() {
```

File: datacapture/src/test/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluatorTest.kt [hunk 6]
```diff
@@ -383,6 +383,110 @@ class EnablementEvaluatorTest {
       .isFalse()
   }
 
+  @Test
+  fun evaluate_greaterThan_answersGreaterThanEnableWhenAnswer_shouldReturnTrue() {
```
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluatorTest.kt:387]:
```suggestion
  fun evaluate_expectAnswerGreaterThanValue_someAnswerGreaterThanValue_shouldReturnTrue() {
```

File: datacapture/src/test/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluatorTest.kt [hunk 7]
```diff
@@ -383,6 +383,110 @@ class EnablementEvaluatorTest {
...
+          operator = Questionnaire.QuestionnaireItemOperator.GREATER_THAN,
+          expected = IntegerType(10),
+          actual = listOf(IntegerType(20))
+        )
+      )
+      .isTrue()
+  }
+
+  @Test
+  fun evaluate_greaterThan_noAnswersGreaterThanEnableWhenAnswer_shouldReturnFalse() {
```
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluatorTest.kt:400]:
```suggestion
  fun evaluate_expectAnswerGreaterThanValue_noAnswerGreaterThanValue_shouldReturnFalse() {
```

File: datacapture/src/test/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluatorTest.kt [hunk 8]
```diff
@@ -383,6 +383,110 @@ class EnablementEvaluatorTest {
...
+          operator = Questionnaire.QuestionnaireItemOperator.GREATER_OR_EQUAL,
+          expected = IntegerType(10),
+          actual = listOf(IntegerType(5))
+        )
+      )
+      .isFalse()
+  }
+
+  @Test
+  fun evaluate_lessThan_answersLessThanEnableWhenAnswer_shouldReturnTrue() {
```
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluatorTest.kt:439]:
```suggestion
  fun evaluate_expectAnswerLessThanValue_someAnswerLessThanValue_shouldReturnTrue() {
```

File: datacapture/src/test/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluatorTest.kt [hunk 9]
```diff
@@ -383,6 +383,110 @@ class EnablementEvaluatorTest {
...
+          operator = Questionnaire.QuestionnaireItemOperator.LESS_THAN,
+          expected = IntegerType(10),
+          actual = listOf(IntegerType(5))
+        )
+      )
+      .isTrue()
+  }
+
+  @Test
+  fun evaluate_lessThan_noAnswersLessThanEnableWhenAnswer_shouldReturnFalse() {
```
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluatorTest.kt:452]:
```suggestion
  fun evaluate_expectAnswerLessThanValue_noAnswerLessThanValue_shouldReturnFalse() {
```

File: datacapture/src/test/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluatorTest.kt [hunk 10]
```diff
@@ -383,6 +383,110 @@ class EnablementEvaluatorTest {
       .isFalse()
   }
 
+  @Test
```
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluatorTest.kt:386]: all these new tests should be moved to line 183.

**Review by jingtang10 — CHANGES_REQUESTED** [review]

**Comment by santosh-pingle** [comment]
Updating code as per review comments.

**Review by jingtang10 — APPROVED** [review]
Thanks @santosh-pingle!

**Pull request merged**